adapters/sharethrough/utils.go

```diff
@@ -0,0 +1,202 @@
+	isAndroid(string) bool
+	isiOS(string) bool
+	isAtMinChromeVersion(string, *regexp.Regexp) bool
+	isAtMinSafariVersion(string, *regexp.Regexp) bool
+}
+
+type Util struct{}
+
+func (u Util) getAdMarkup(strResp openrtb_ext.ExtImpSharethroughResponse, params *StrAdSeverParams) (string, error) {
+	strRespId := fmt.Sprintf("str_response_%s", strResp.BidID)
+	jsonPayload, err := json.Marshal(strResp)
+	if err != nil {
+		return "", err
+	}
+
+	tmplBody := `
+		<img src="//b.sharethrough.com/butler?type=s2s-win&arid={{.Arid}}" />
+
+		<div data-str-native-key="{{.Pkey}}" data-stx-response-name="{{.StrRespId}}"></div>
+	 	<script>var {{.StrRespId}} = "{{.B64EncodedJson}}"</script>
+	`
+
+	if params.Iframe {
+		tmplBody = tmplBody + `
+			<script src="//native.sharethrough.com/assets/sfp.js"></script>
+		`
+	} else {
+		tmplBody = tmplBody + `
+			<script src="//native.sharethrough.com/assets/sfp-set-targeting.js"></script>
+	    	<script>
+		     (function() {
+		       if (!(window.STR && window.STR.Tag) && !(window.top.STR && window.top.STR.Tag)) {
+		         var sfp_js = document.createElement('script');
+		         sfp_js.src = "//native.sharethrough.com/assets/sfp.js";
+		         sfp_js.type = 'text/javascript';
+		         sfp_js.charset = 'utf-8';
+		         try {
+		             window.top.document.getElementsByTagName('body')[0].appendChild(sfp_js);
+		         } catch (e) {
+		           console.log(e);
+		         }
+		       }
+		     })()
+		   </script>
+		`
+	}
+
+	tmpl, err := template.New("sfpjs").Parse(tmplBody)
+	if err != nil {
+		return "", err
+	}
+
+	var buf []byte
+	templatedBuf := bytes.NewBuffer(buf)
+
+	b64EncodedJson := base64.StdEncoding.EncodeToString(jsonPayload)
+	err = tmpl.Execute(templatedBuf, struct {
+		Arid           template.JS
+		Pkey           string
+		StrRespId      template.JS
+		B64EncodedJson string
+	}{
+		template.JS(strResp.AdServerRequestID),
+		params.Pkey,
+		template.JS(strRespId),
+		b64EncodedJson,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	return templatedBuf.String(), nil
+}
+
+func (u Util) getPlacementSize(formats []openrtb.Format) (height uint64, width uint64) {
+	biggest := struct {
+		Height uint64
+		Width  uint64
+	}{
+		Height: 1,
+		Width:  1,
+	}
+
+	for i := 0; i < len(formats); i++ {
+		format := formats[i]
+		if (format.H * format.W) > (biggest.Height * biggest.Width) {
+			biggest.Height = format.H
+			biggest.Width = format.W
+		}
+	}
+
+	return biggest.Height, biggest.Width
+}
+
+func (u Util) canAutoPlayVideo(userAgent string, parsers UserAgentParsers) bool {
+	if u.isAndroid(userAgent) {
+		if u.isAtMinChromeVersion(userAgent, parsers.ChromeVersion) {
+			return true
+		} else {
+			return false
+		}
+	} else if u.isiOS(userAgent) {
+		if u.isAtMinSafariVersion(userAgent, parsers.SafariVersion) || u.isAtMinChromeVersion(userAgent, parsers.ChromeiOSVersion) {
+			return true
+		} else {
+			return false
+		}
+	} else {
+		return true
+	}
+}
```

Comment:
Team, are we sure the logic inside the refactored function equals the original boolean statement? From Joe's comment last Friday, the original logic in `canAutoPlayVideo(userAgent string)` was:
```
+func (u Util) canAutoPlayVideo(userAgent string) bool {
+	return (u.isAndroid(userAgent) && u.isAtMinChromeVersion(userAgent)) ||
+		(u.isiOS(userAgent) &&
+			(u.isAtMinSafariVersion(userAgent) || u.isAtMinChromeIosVersion(userAgent))) ||
+		!(u.isAndroid(userAgent) || u.isiOS(userAgent))
```
Said return statement, unless I got it wrong, translates to the expression in line 44 below. The program that follows compares both the old and new versions of this function and they don't seem to return the same results in scenarios 25, 26 and 27 of our truth table.
```
  1 package main
  2 import "fmt"
  3 type testCase struct {
  4     isAndroid, isiOS, isAtMinChromeVersion, isAtMinSafariVersion, isAtMinChromeIosVersion bool
  5 }
  6 func main() {
  7     var tests []testCase = []testCase{
  8         {false, false, false, false, false},
  9         {false, false, false, false, true},
 10         {false, false, false, true, false},
 11         {false, false, false, true, true},
 12         {false, false, true, false, false},
 13         {false, false, true, false, true},
 14         {false, false, true, true, false},
 15         {false, false, true, true, true},
 16         {false, true, false, false, false},
 17         {false, true, false, false, true},
 18         {false, true, false, true, false},
 19         {false, true, false, true, true},
 20         {false, true, true, false, false},
 21         {false, true, true, false, true},
 22         {false, true, true, true, false},
 23         {false, true, true, true, true},
 24         {true, false, false, false, false},
 25         {true, false, false, false, true},
 26         {true, false, false, true, false},
 27         {true, false, false, true, true},
 28         {true, false, true, false, false},
 29         {true, false, true, false, true},
 30         {true, false, true, true, false},
 31         {true, false, true, true, true},
 32         {true, true, false, false, false},
 33         {true, true, false, false, true},
 34         {true, true, false, true, false},
 35         {true, true, false, true, true},
 36         {true, true, true, false, false},
 37         {true, true, true, false, true},
 38         {true, true, true, true, false},
 39         {true, true, true, true, true},
 40     }
 41     fmt.Println("#\tisAndroid\t isiOS \tisAtMinChromeVersion\tisAtMinSafariVersion \tisAtMinChromeIosVersion ")
 42     for i := 0; i < len(tests); i++ {
 43         var msg string
 44         res := (tests[i].isAndroid && tests[i].isAtMinChromeVersion) || (tests[i].isiOS && (tests[i].isAtMinSafariVersion || tests[i].isAtMinChromeIosVersion)) || (!tests[i].isAndroid && !tests[i].isiOS)
 45         if res == approvedLogic(tests[i]) {
 46             msg = "passed"
 47         } else {
 48             msg = "NOT PASSED"
 49         }
 50         fmt.Printf("%d\t %t \t\t%t\t\t %t \t\t %t \t\t\t\t %t \t\t\t%s\n", i, tests[i].isAndroid, tests[i].isiOS, tests[i].isAtMinChromeVersion, tests[i].isAtMinSafariVersion, tests[i].isAtMinChromeIosVersion, msg)
 51     }
 52 }
 53 func approvedLogic(test testCase) bool {
 54     if test.isAndroid {
 55         if test.isAtMinChromeVersion {
 56             return true
 57         } else {
 58             return false
 59         }
 60     } else if test.isiOS {
 61         if test.isAtMinSafariVersion || test.isAtMinChromeIosVersion {
 62             return true
 63         } else {
 64             return false
 65         }
 66     } else {
 67         return true
 68     }
 69 }
```
## Test case results:
```
╰─➤  go run booleanTest2.go
#       isAndroid        isiOS  isAtMinChromeVersion    isAtMinSafariVersion    isAtMinChromeIosVersion
0        false          false            false           false                           false                  passed
1        false          false            false           false                           true                   passed
2        false          false            false           true                            false                  passed
3        false          false            false           true                            true                   passed
4        false          false            true            false                           false                  passed
5        false          false            true            false                           true                   passed
6        false          false            true            true                            false                  passed
7        false          false            true            true                            true                   passed
8        false          true             false           false                           false                  passed
9        false          true             false           false                           true                   passed
10       false          true             false           true                            false                  passed
11       false          true             false           true                            true                   passed
12       false          true             true            false                           false                  passed
13       false          true             true            false                           true                   passed
14       false          true             true            true                            false                  passed
15       false          true             true            true                            true                   passed
16       true           false            false           false                           false                  passed
17       true           false            false           false                           true                   passed
18       true           false            false           true                            false                  passed
19       true           false            false           true                            true                   passed
20       true           false            true            false                           false                  passed
21       true           false            true            false                           true                   passed
22       true           false            true            true                            false                  passed
23       true           false            true            true                            true                   passed
24       true           true             false           false                           false                  passed
25       true           true             false           false                           true                   NOT PASSED
26       true           true             false           true                            false                  NOT PASSED
27       true           true             false           true                            true                   NOT PASSED
28       true           true             true            false                           false                  passed
29       true           true             true            false                           true                   passed
30       true           true             true            true                            false                  passed
31       true           true             true            true                            true                   passed
```

Reply:
The following proposed logic fits the original function better as it passes all test cases:
```
func rearangedLogic(test testCase) bool {
	if test.isAndroid {
		return test.isAtMinChromeVersion || (test.isiOS && (test.isAtMinSafariVersion || test.isAtMinChromeIosVersion))
	}
	if test.isiOS && !test.isAtMinSafariVersion && !test.isAtMinChromeIosVersion {
		return false
	}
	return true
}
```

Reply:
Or this one:
```
func rearangedLogicV2(test testCase) bool {
	if test.isAndroid {
		return test.isAtMinChromeVersion || (test.isiOS && (test.isAtMinSafariVersion || test.isAtMinChromeIosVersion))
	}
	return !test.isiOS || test.isAtMinSafariVersion || test.isAtMinChromeIosVersion
}
```

Reply:
@guscarreon thanks for deeply double checking that! confirming that's not an issue, `isAndroid` and `isiOS` are mutually exclusive, so you could omit all the `true/true` test cases here. Also even if for some reason we would have both true at the same time, the newest version is the correct one.

Reply:
Awesome @maphe. Without changing the logic, do you think it is worth to refactor the function a bit so we have less exit points?
```
func (u Util) canAutoPlayVideo(userAgent string, parsers UserAgentParsers) bool {
	if u.isAndroid(userAgent) {
		return u.isAtMinChromeVersion(userAgent, parsers.ChromeVersion) {
	} else if u.isiOS(userAgent) {
		return u.isAtMinSafariVersion(userAgent, parsers.SafariVersion) || u.isAtMinChromeVersion(userAgent, parsers.ChromeiOSVersion) {
	}
	return true
}
```